### DISCUSSION TITLE
Upload image data directly from FlaskForm into pdf without uploading it to the server.

### DISCUSSION THREAD

**Discussion by megaprimatus** [py-pdf/fpdf2#599] [Q&A]
Hi,
I'm trying to figure out how I can place image data to pdf document without uploading it to the server. 

In other words, I have a form (FlaskForm) in the page where the user fills in some data and one field is FileField. I use that data to place it into PDF document. String fields works fine but image.data doesn't want to work. It will probably work if I first save the image in the server after form submission and then specify the path to that image. But I really don't want to save the image to the server but use its data directly to upload it in the PDF. Here is my code:

```
class ExportForm(FlaskForm):
    project_name = StringField('Name of the  project')
    doc_number = StringField('Document of the report')
    designer = StringField('Created by')
    logo = FileField('Upload your logo', validators=[FileAllowed(['jpg', 'png'], 'not allowed')])
    export = SubmitField('export')

if export_form.export.data and export_form.validate_on_submit():
    class PDF(FPDF, HTMLMixin):
        def header(self):               
            self.cell(0, 10, export_form.project_name.data, 0, 1, 'L')               
            self.cell(0, 5, export_form.doc_number.data, 0, 1, 'L')               
            self.cell(0, 5, export_form.designer.data, 0, 1, 'L')       
            self.image('export_form.logo.data', 175, 10, 25, 0)
```

**Comment by gmischler** [comment]
> I'm trying to figure out how I can place image data to pdf document without uploading it to the server.

The docstring for the `image()` method includes the following:

    def image(self, name, x=None, y=None, w=0, h=0, type='', link='', title=None, alt_text=None)
        Put an image on the page.
        ...
        Args

        name
            either a string representing a file path to an image,
            an URL to an image,
            an io.BytesIO,
            or a instance of PIL.Image.Image

I suspect your image data comes in as a bytes object. You now have a choice of converting that either to a `io.BytesIO()` object, or a `PIL.Image.Image()` object, either of which will be accepted by `image()`.

I'm not sure why simply supplying a bytes object is not supported. That may have historical reasons from a time when byte and text strings weren't as clearly distinguished as they are today. The API of using the same parameter for both file names and actual data also doesn't provide optimal flexibility.
If anyone can come up with a more versatile API concept here, which can be implemented while maintaining backwards compatibility, suggestions or a PR are welcome.

**Reply by megaprimatus** [reply]
Thanks for the answer. I did search yesterday and found that Flask stores the image data in this `request.files['name_of_FileField'] `
So I simply changed it and it worked:
` self.image('request.files['logo']', 175, 10, 25, 0)`

**Reply by gmischler** [reply]
The contents of `request.files` appear to be [`werkzeug.FileStorage()`](https://tedboy.github.io/flask/generated/generated/werkzeug.FileStorage.html#werkzeug.FileStorage) instances. Those again seem to mimick `io.BytesIO()` well enough to be processed by fpdf2 as such.

Good to know!
